test/integration/retryable-reads/retryable_reads.spec.prose.test.ts

```diff
@@ -0,0 +1,146 @@
+  let client: MongoClient, failPointName;
+
+  afterEach(async () => {
+    try {
+      if (failPointName) {
+        await client.db('admin').command({ configureFailPoint: failPointName, mode: 'off' });
+      }
+    } finally {
+      failPointName = undefined;
+      await client?.close();
+    }
+  });
+
+  describe('PoolClearedError Retryability Test', () => {
+    // This test will be used to ensure drivers properly retry after encountering PoolClearedErrors.
+    // It MUST be implemented by any driver that implements the CMAP specification.
+    // This test requires MongoDB 4.2.9+ for blockConnection support in the failpoint.
+
+    let observedEvents: Array<{ name: string; event: Record<string, any> }>;
+    let testCollection: Collection;
+    beforeEach(async function () {
+      // 1. Create a client with maxPoolSize=1 and retryReads=true.
+      // If testing against a sharded deployment, be sure to connect to only a single mongos. <-- TODO: what does that look like?
+      client = this.configuration.newClient({
+        maxPoolSize: 1,
+        retryReads: true,
+        monitorCommands: true
+      });
+      await client.connect();
+
+      testCollection = client.db('retryable-reads-prose').collection('pool-clear-retry');
+      await testCollection.drop().catch(() => null);
+      await testCollection.insertMany([{ test: 1 }, { test: 2 }]);
+
+      // 2. Enable the following failpoint:
+      // NOTE: "9. Disable the failpoint" is done in afterEach
+      failPointName = 'failCommand';
+      const failPoint = await client.db('admin').command({
+        configureFailPoint: failPointName,
+        mode: { times: 1 },
+        data: {
+          failCommands: ['find'],
+          errorCode: 91,
+          blockConnection: true,
+          blockTimeMS: 1000
+        }
+      });
+
+      expect(failPoint).to.have.property('ok', 1);
+
+      observedEvents = [];
+      for (const observedEvent of [
+        'connectionCheckOutStarted',
+        'connectionCheckedOut',
+        'connectionCheckOutFailed',
+        'connectionPoolCleared',
+        'commandStarted'
+      ]) {
+        client.on(observedEvent, ev => {
+          observedEvents.push({ name: observedEvent, event: ev });
+        });
+      }
+    });
+
+    it('should emit events in the expected sequence', {
+      metadata: { requires: { mongodb: '>=4.2.9', topology: '!load-balanced' } },
+      test: async function () {
+        // 3. Start two threads and attempt to perform a findOne simultaneously on both.
+        const results = await Promise.all([
+          testCollection.findOne({ test: 1 }),
+          testCollection.findOne({ test: 2 })
+        ]);
+
+        client.removeAllListeners();
+        // 4. Verify that both findOne attempts succeed.
+        expect(results[0]).to.have.property('test', 1);
+        expect(results[1]).to.have.property('test', 2);
+
+        // 5. Via CMAP monitoring, assert that the first check out succeeds.
```

Comment:
For this sequence I found if the observed events were split into 2 arrays for cmap events and command events, I could reduce all the code for asserting the events down to this:

```js
// 5. Via CMAP monitoring, assert that the first check out succeeds.
expect(cmapEvents.shift().name).to.equal('connectionCheckOutStarted', 'expected 1) checkout 1 to start');
expect(cmapEvents.shift().name).to.equal('connectionCheckOutStarted', 'expected 2) checkout 2 to start');
expect(cmapEvents.shift().name).to.equal('connectionCheckedOut', 'expected 3) first checkout to succeed');
// 6. Via CMAP monitoring, assert that a PoolClearedEvent is then emitted.
expect(cmapEvents.shift().name).to.equal('connectionPoolCleared', 'expected 4) pool to clear');
// 7. Via CMAP monitoring, assert that the second check out then fails due
// to a connection error.
const nextEvent = cmapEvents.shift();
expect(nextEvent.name).to.equal('connectionCheckOutFailed', 'expected 5) checkout 2 to fail');
expect(nextEvent.event).to.have.property('reason', 'connectionError');

// 8. Via Command Monitoring, assert that exactly three find CommandStartedEvents were observed in total.
expect(commandEvents).to.have.lengthOf(3, 'expected 3 find command started events');
```

It's obviously less code but curious if there was a specific reason for doing the find by index method?

Reply:
My original reasoning for using indices was because I did not want to make assumptions about any other monitoring/pool events that may or may not happen in between the events of interest, but I suppose since we explicitly opt into listening for explicit events, asserting on the exact sequence of the tracked events is relatively low risk for us even if the future implementation of cmap or monitoring changes to emit additional events. What are your thoughts?